plugins/rts54hid/fu-rts54hid-device.c

```diff
@@ -0,0 +1,396 @@
+		.bufferlen = GUINT16_TO_LE (data_sz),
+		.parameters = 0,
+	};
+	guint8 buf[FU_RTS54HID_REPORT_LENGTH] = { 0 };
+
+	g_return_val_if_fail (data_sz <= 128, FALSE);
+	g_return_val_if_fail (data != NULL, FALSE);
+	g_return_val_if_fail (data_sz != 0, FALSE);
+
+	memcpy (buf, &cmd_buffer, sizeof(cmd_buffer));
+	memcpy (buf + FU_RTS54HID_CMD_BUFFER_OFFSET_DATA, data, data_sz);
+	if (!fu_rts54hid_device_set_report (self, buf, sizeof(buf), error)) {
+		g_prefix_error (error, "failed to write flash @%08x: ", (guint) addr);
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_rts54hid_device_verify_update_fw (FuRts54HidDevice *self, GError **error)
+{
+	const FuRts54HidCmdBuffer cmd_buffer = {
+		.cmd = FU_RTS54HID_CMD_WRITE_DATA,
+		.ext = FU_RTS54HID_EXT_VERIFYUPDATE,
+		.cmd_data0 = 1,
+		.cmd_data1 = 0,
+		.cmd_data2 = 0,
+		.cmd_data3 = 0,
+		.bufferlen = GUINT16_TO_LE (1),
+		.parameters = 0,
+	};
+	guint8 buf[FU_RTS54HID_REPORT_LENGTH] = { 0 };
+
+	/* set then get */
+	memcpy (buf, &cmd_buffer, sizeof(cmd_buffer));
+	if (!fu_rts54hid_device_set_report (self, buf, sizeof(buf), error))
+		return FALSE;
+	g_usleep (4 * G_USEC_PER_SEC);
+	if (!fu_rts54hid_device_get_report (self, buf, sizeof(buf), error))
+		return FALSE;
+
+	/* check device status */
+	if (buf[0x40] != 0x01) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_WRITE,
+				     "firmware flash failed");
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_rts54hid_device_erase_spare_bank (FuRts54HidDevice *self, GError **error)
+{
+	FuRts54HidCmdBuffer cmd_buffer = {
+		.cmd = FU_RTS54HID_CMD_WRITE_DATA,
+		.ext = FU_RTS54HID_EXT_ERASEBANK,
+		.cmd_data0 = 0,
+		.cmd_data1 = 1,
+		.cmd_data2 = 0,
+		.cmd_data3 = 0,
+		.bufferlen = 0,
+		.parameters = 0,
+	};
+	guint8 buf[FU_RTS54HID_REPORT_LENGTH] = { 0 };
+	memcpy (buf, &cmd_buffer, sizeof(cmd_buffer));
+	if (!fu_rts54hid_device_set_report (self, buf, sizeof(buf), error)) {
+		g_prefix_error (error, "failed to erase spare bank: ");
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_rts54hid_device_ensure_status (FuRts54HidDevice *self, GError **error)
+{
+	const FuRts54HidCmdBuffer cmd_buffer = {
+		.cmd = FU_RTS54HID_CMD_READ_DATA,
+		.ext = FU_RTS54HID_EXT_READ_STATUS,
+		.cmd_data0 = 0,
+		.cmd_data1 = 0,
+		.cmd_data2 = 0,
+		.cmd_data3 = 0,
+		.bufferlen = GUINT16_TO_LE (32),
+		.parameters = 0,
+	};
+	guint8 buf[FU_RTS54HID_REPORT_LENGTH] = { 0 };
+	g_autofree gchar *version = NULL;
+
+	/* set then get */
+	memcpy (buf, &cmd_buffer, sizeof(cmd_buffer));
+	if (!fu_rts54hid_device_set_report (self, buf, sizeof(buf), error))
+		return FALSE;
+	if (!fu_rts54hid_device_get_report (self, buf, sizeof(buf), error))
+		return FALSE;
+
+	/* check the hardware capabilities */
+	self->dual_bank = (buf[0x40 + 7] & 0xf0) == 0x80;
+	self->fw_auth = (buf[0x40 + 13] & 0x02) > 0;
+	self->status1 = buf[0x40 + 25];
+	self->status2 = buf[0x40 + 26];
```

Comment:
Maybe a single guint16 called `status` then?

Reply:
I think that makes sense.  The point I was trying to relay (but failed) is that it may be better to have a separate function for `update_status` that reads this and saves it.  Reading/checking the other bits again and updating the version again probably are not necessary however.

Reply:
status[25,26] only for Dell Dock to update EC fw, the 2 byte not for update Realtek Hub fw pass through HID, only Dell Dock update fw need those 2 byte

Reply:
Then I'll remove them -- thanks!